### PR TITLE
Add conf.nameservers + small dns resolver

### PR DESCRIPTION
**Pull request by gpotter2**
- adds `conf.nameservers` which is read from either `resolv.conf` or the windows C API
- adds a small `dns_resolve` (propose better name?) util that queries the server in `conf.nameservers`
- fixes issues with `CacheInstance` and minor cleanups (found randomly while doing this)

The purpose of this is to have a very simple and reliable API to use in other Scapy scripts / functions, and not to have the more "craft-it-yourself" approach (that's also very nice) you'd have with `sr1(IP()/UDP()/DNS())`. I'm using that in a future PR.

### CODE REVIEW

File: scapy/layers/dns.py [hunk 1]
```diff
@@ -1046,6 +1051,55 @@ def pre_dissect(self, s):
...
+
+    kwargs.setdefault("timeout", 5)
+    kwargs.setdefault("verbose", 0)
+    for nameserver in conf.nameservers:
+        try:
+            sock = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
+            sock.connect((nameserver, 53))
+            sock = StreamSocket(sock, DNS)
+            res = sock.sr1(
+                DNS(qd=[DNSQR(qname=qname, qtype=qtype)]),
```
Comment (guedou) [inline scapy/layers/dns.py:1076]: We need to set the `DNS.id` field to a random value. As Scapy always uses 0 (`DNS(raw(DNS())).id == DNS(raw(DNS())).id == 0`), that could lead to some attacks.

File: scapy/layers/dns.py [hunk 2]
```diff
@@ -1046,6 +1051,55 @@ def pre_dissect(self, s):
...
+                if verbose:
+                    log_runtime.info(
+                        "DNS: %s answered with failure for %s" % (
+                            nameserver,
+                            qname,
+                        )
+                    )
+            else:
+                break
+    if res is not None and res.an:
```
Comment (guedou) [inline scapy/layers/dns.py:1097]: you should also check that the qname and qtypes match the answer name and types. also, i don't know if it is safe to assume that the answer will always be in the first `an` element.

**Review by guedou — CHANGES_REQUESTED** [review]

File: scapy/arch/windows/__init__.py [hunk 3]
```diff
@@ -1016,3 +1024,16 @@ def __init__(self, *args, **kargs):
...
+
+#######
+# DNS #
+#######
+
+def read_nameservers() -> List[str]:
+    """Return the nameservers configured by the OS (on the default interface)
+    """
+    # Windows has support for different DNS servers on each network interface,
+    # which is kinda dope... but linux is dumb. So this function is just an
```
[the remark this answers is not in the document]
Comment (gpotter2) [reply scapy/arch/windows/__init__.py:1037]: Heh, right !

File: scapy/arch/unix.py [hunk 4]
```diff
@@ -409,3 +410,18 @@ def read_routes6():
 
     fd_netstat.close()
     return routes
+
+
+#######
+# DNS #
+#######
+
+def read_nameservers() -> List[str]:
```
[the remark this answers is not in the document]
Comment (gpotter2) [reply scapy/arch/unix.py:419]: Maybe, but seems a bit out of scope for this PR. I'm not sure it has an equivalence in the Windows world.

File: scapy/layers/dns.py (hunk 1 above)
Comment (gpotter2) [reply scapy/layers/dns.py:1076]: Good call !

File: scapy/layers/dns.py (hunk 2 above)
Comment (gpotter2) [reply scapy/layers/dns.py:1097]: Good call ! And you're right we should check them all.

File: scapy/arch/unix.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (guedou) [reply scapy/arch/unix.py:419]: I am Ok with the furent implementation.

**Comment by gpotter2** [comment]
Great, feel free to merge if current implementation looks ok.

**Pull request merged**